pex/pex_builder.py

```diff
@@ -495,17 +495,15 @@ def add_dist_location(
         self.add_requirement(distribution.as_requirement())
 
     def _precompile_source(self):
+        vendored_dir = os.path.join(self._pex_info.bootstrap, "pex/vendor/_vendored")
         source_relpaths = [
             path
             for label in ("source", "executable", "main", "bootstrap")
             for path in self._chroot.filesets.get(label, ())
             if path.endswith(".py")
-            # N.B.: This file if Python 3.6+ only and will not compile under Python 2.7 or
-            # Python 3.5. Since we don't actually use it we just skip compiling it.
-            and path
-            != os.path.join(
-                self._pex_info.bootstrap, "pex/vendor/_vendored/attrs/attr/_next_gen.py"
-            )
+            # N.B.: Some of our vendored code does not work with all versions of Python we support;
```

Comment:
Previously we were excluding a single file, now we're excluding all the vendored boostrap code en masse. Will this not be a perf hit that belies the promise of bytecode_compile=True? 

Would it make sense (in a followup, if at all) to best-effort this and ignore any errors? Or is the performance hit all in the noise anyway? The code in _COMPILER_MAIN already splits out successes and errors, and it would just mean propagating that to the invoker.

Reply:
It had become irrelevant. The only time the bootstrap code is executed from within a zipped PEX is on 1st cold start, after that the PEX runs from either a PEX_ROOT venv or a PEX_ROOT unzipped_pexes directory. In either of those cases pycs are compiled and stick around. As such microoptimizing the known slow case of a cold start just didn't seem close to worth it.

Reply:
Makes sense!

Reply:
Ok, the colde-start is surprisingly bad - I'll dig in to see why before landing this:
```
$ pex -oold-empty.pex
$ pex --compile -oold-empty-compiled.pex
$ python -mpex -onew-empty.pex
$ python -mpex --compile -onew-empty-compiled.pex

$ hyperfine -p 'rm -rf ~/.pex' './old-empty.pex -c ""' './old-empty-compiled.pex -c ""' './new-empty.pex -c ""' './new-empty-compiled.pex -c ""'
Benchmark 1: ./old-empty.pex -c ""
  Time (mean ± σ):     259.5 ms ±   3.5 ms    [User: 235.1 ms, System: 25.6 ms]
  Range (min … max):   253.3 ms … 262.7 ms    10 runs

Benchmark 2: ./old-empty-compiled.pex -c ""
  Time (mean ± σ):     277.8 ms ±  15.7 ms    [User: 249.7 ms, System: 27.8 ms]
  Range (min … max):   265.4 ms … 317.5 ms    11 runs

Benchmark 3: ./new-empty.pex -c ""
  Time (mean ± σ):     411.5 ms ±   8.3 ms    [User: 381.9 ms, System: 30.7 ms]
  Range (min … max):   403.3 ms … 431.6 ms    10 runs

Benchmark 4: ./new-empty-compiled.pex -c ""
  Time (mean ± σ):     423.0 ms ±   6.5 ms    [User: 395.0 ms, System: 29.0 ms]
  Range (min … max):   413.9 ms … 433.5 ms    10 runs

Summary
  './old-empty.pex -c ""' ran
    1.07 ± 0.06 times faster than './old-empty-compiled.pex -c ""'
    1.59 ± 0.04 times faster than './new-empty.pex -c ""'
    1.63 ± 0.03 times faster than './new-empty-compiled.pex -c ""'

$ hyperfine -w 2 './old-empty.pex -c ""' './old-empty-compiled.pex -c ""' './new-empty.pex -c ""' './new-empty-compiled.pex -c ""'
Benchmark 1: ./old-empty.pex -c ""
  Time (mean ± σ):     170.6 ms ±   3.6 ms    [User: 158.6 ms, System: 13.3 ms]
  Range (min … max):   164.7 ms … 175.6 ms    17 runs

Benchmark 2: ./old-empty-compiled.pex -c ""
  Time (mean ± σ):     172.1 ms ±   3.3 ms    [User: 157.8 ms, System: 15.6 ms]
  Range (min … max):   167.0 ms … 179.5 ms    17 runs

Benchmark 3: ./new-empty.pex -c ""
  Time (mean ± σ):     170.8 ms ±   3.6 ms    [User: 158.3 ms, System: 13.8 ms]
  Range (min … max):   166.2 ms … 176.5 ms    17 runs

Benchmark 4: ./new-empty-compiled.pex -c ""
  Time (mean ± σ):     174.2 ms ±   3.7 ms    [User: 157.6 ms, System: 17.9 ms]
  Range (min … max):   167.8 ms … 178.6 ms    17 runs

Summary
  './old-empty.pex -c ""' ran
    1.00 ± 0.03 times faster than './new-empty.pex -c ""'
    1.01 ± 0.03 times faster than './old-empty-compiled.pex -c ""'
    1.02 ± 0.03 times faster than './new-empty-compiled.pex -c ""'
```